File: day02/main.py
#! /usr/bin/env python3
import re
import itertools


def validateA(d):
    n = d["password"].count(d["ch"])
    return n >= int(d["ix1"]) and n <= int(d["ix2"])


def validateB(d):
    ix1 = int(d["ix1"]) - 1
    ix2 = int(d["ix2"]) - 1
    at1 = ix1 < len(d["password"]) and d["password"][ix1 : ix1 + 1] == d["ch"]
    at2 = ix2 < len(d["password"]) and d["password"][ix2 : ix2 + 1] == d["ch"]
    return at1 != at2


if __name__ == "__main__":
    f = open("./input.txt")
    lines = f.readlines()
    pat = re.compile(
        "(?P<ix1>[0-9]+)-(?P<ix2>[0-9]+) (?P<ch>[a-z]): (?P<password>[a-z]+)"
    )
    data = [pat.match(l).groupdict() for l in lines]
    valids = [x for x in data if validateA(x)]
    print(len(valids))
    valids2 = [x for x in data if validateB(x)]
    print(len(valids2))

File: day02/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// PWData holds the password data as specified in the problem set
type PWData struct {
	Letter   string
	MinCount int
	MaxCount int
	Password string
}

// Validate checks if the pwdata meets the validation rules
func (p PWData) Validate() bool {
	if p.Letter == "" {
		return false
	}
	n := strings.Count(p.Password, p.Letter)
	return n >= p.MinCount && n <= p.MaxCount
}

func parsePassword(s string) PWData {
	pat := regexp.MustCompile("([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)")
	matches := pat.FindStringSubmatch(s)
	if len(matches) != 5 {
		log.Fatalf("line: %s does not match pattern! %s", s, matches)
	}
	mincount, _ := strconv.Atoi(matches[1])
	maxcount, _ := strconv.Atoi(matches[2])
	return PWData{
		Letter:   matches[3],
		MinCount: mincount,
		MaxCount: maxcount,
		Password: matches[4],
	}
}

func day02a(lines []string) {
	nvalid := 0
	for _, line := range lines {
		pwd := parsePassword(line)
		if pwd.Validate() {
			nvalid++
		}
	}
	fmt.Printf("%d passwords are valid\n", nvalid)
}

func main() {
	f, err := os.Open("./input.txt")
	if err != nil {
		log.Fatal(err)
	}
	b, err := ioutil.ReadAll(f)
	if err != nil {
		log.Fatal(err)
	}
	lines := strings.Split(string(b), "\n")

	day02a(lines)
}
